libs/language-server/src/lib/ast/wrappers/wrapper-factory.ts

```diff
@@ -0,0 +1,294 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { strict as assert } from 'assert';
+
+import {
+  AstNode,
+  type Reference,
+  assertUnreachable,
+  isReference,
+} from 'langium';
+
+import { type OperatorEvaluatorRegistry } from '../expressions';
+import {
+  BlocktypePipeline,
+  BuiltinConstrainttypeDefinition,
+  CellRangeLiteral,
+  CompositeBlocktypeDefinition,
+  PipeDefinition,
+  PipelineDefinition,
+  type ReferenceableBlocktypeDefinition,
+  isBuiltinConstrainttypeDefinition,
+  isReferenceableBlocktypeDefinition,
+} from '../generated/ast';
+
+import { AstNodeWrapper } from './ast-node-wrapper';
+import { CellRangeWrapper } from './cell-range-wrapper';
+import { PipeWrapper } from './pipe-wrapper';
+import { PipelineWrapper } from './pipeline-wrapper';
+// eslint-disable-next-line import/no-cycle
+import { BlockTypeWrapper } from './typed-object/blocktype-wrapper';
+import { ConstraintTypeWrapper } from './typed-object/constrainttype-wrapper';
+
+abstract class AstNodeWrapperFactory<
+  N extends AstNode,
+  W extends AstNodeWrapper<N>,
+> {
+  abstract canWrap(toBeWrapped: N | Reference<N>): boolean;
+  abstract doWrap(toBeWrapped: N | Reference<N>): W;
+
+  wrap(toBeWrapped: N | Reference<N>): W {
+    assert(
+      this.canWrap(toBeWrapped),
+      `AstNode ${this.getName(toBeWrapped)} cannot be wrapped`,
+    );
+    return this.doWrap(toBeWrapped);
+  }
+
+  private getName(toBeWrapped: N | Reference<N>): string {
+    const node = isReference(toBeWrapped) ? toBeWrapped.ref : toBeWrapped;
+    if (node === undefined) {
+      return '<unresolved reference>';
+    }
+    if ('name' in node && typeof node.name === 'string') {
+      return node.name;
+    }
+    return '<unnamed ast node>';
+  }
+}
+
+export class WrapperFactory {
```

Comment:
Naming is a hard thing, and I think it makes sense to spend some time thinking about it.

In the current design, the "factory container" is more used as a namespacing mechanism (which might be discussable whether that is appropriate or not). So the access pattern currently is `<factories object>.<namespace>.<method>`. From your comment, I assume that especially the name of the factory container object (and its class) is the thing under discussion, but the other both components are fine?

If that is the case, then I also like `wrapperFactories` instead of `wrapperFactory` as parameter/variable names.
For the class, we could call it `WrapperFactoryContainer`, `WrapperFactoryService`, or `WrapperFactoryProvider` instead of `WrapperFactory`? Personally, I like `WrapperFactoryProvider`  most: it provides the factories, and the naming blends in will with the other Langium services.

Reply:
> From your comment, I assume that especially the name of the factory container object (and its class) is the thing under discussion, but the other both components are fine?

What are the other two components?

Reply:
I like `WrapperFactoryProvider` best as well.

Reply:
> > From your comment, I assume that especially the name of the factory container object (and its class) is the thing under discussion, but the other both components are fine?
> 
> What are the other two components?

namespace and method

Reply:
Implemented in [b167c38](https://github.com/jvalue/jayvee/pull/536/commits/b167c38342a6cc826b4ba6c42b7f80d4ed7c0c80)